algorithms/src/snark/marlin/ahp/matrices.rs

```diff
@@ -124,81 +120,67 @@ pub struct MatrixArithmetization<F: PrimeField> {
-}
-
 pub(crate) fn matrix_evals<F: PrimeField>(
     matrix: &Matrix<F>,
     non_zero_domain: &EvaluationDomain<F>,
-    constraint_domain: &EvaluationDomain<F>,
+    variable_domain: &EvaluationDomain<F>,
     input_domain: &EvaluationDomain<F>,
-    elems: &[F],
-    eq_poly_vals: &HashMap<F, F>,
-) -> MatrixEvals<F> {
+    constraint_domain_elems: &[F],
+    variable_domain_elems: &[F],
+) -> Result<MatrixEvals<F>, anyhow::Error> {
     let lde_evals_time = start_timer!(|| "Computing row, col and val evals");
 
-    // Recall that we are computing the arithmetization of M^*,
-    // where `M^*(i, j) := M(j, i) * u_H(j, j)`.
+    // We are computing the arithmetization of M^,
+    // where `M^(α,β) = \sum_{κ∈K} val(κ)·L^R_row(κ)(α)·L^C_col(κ)(β)`
 
-    let mut row_vec = Vec::with_capacity(non_zero_domain.size());
-    let mut col_vec = Vec::with_capacity(non_zero_domain.size());
-    let mut val_vec = Vec::with_capacity(non_zero_domain.size());
-    let mut inverses = Vec::with_capacity(non_zero_domain.size());
+    let mut row_indices = Vec::with_capacity(non_zero_domain.size());
+    let mut col_indices = Vec::with_capacity(non_zero_domain.size());
+    let mut row_col_vals = Vec::with_capacity(non_zero_domain.size());
     let mut count = 0;
 
-    // Recall that we are computing the arithmetization of M^*,
-    // where `M^*(i, j) := M(j, i) * u_H(j, j)`.
-    for (r, row) in matrix.iter().enumerate() {
-        for (val, i) in row {
-            let row_val = elems[r];
-            let col_val = elems[constraint_domain.reindex_by_subdomain(input_domain, *i)];
+    for (row_index, row) in matrix.iter().enumerate() {
+        for (val, input_var_index) in row {
+            let row_i = constraint_domain_elems[row_index];
+            let col_i = variable_domain_elems[variable_domain.reindex_by_subdomain(input_domain, *input_var_index)?];
 
-            // We are dealing with the transpose of M
-            row_vec.push(col_val);
-            col_vec.push(row_val);
-            val_vec.push(*val);
-            inverses.push(eq_poly_vals[&col_val]);
+            row_indices.push(row_i);
+            col_indices.push(col_i);
+            row_col_vals.push(*val);
 
             count += 1;
         }
     }
 
-    batch_inversion::<F>(&mut inverses);
+    let mut row_col_indices: Vec<_> = row_indices.iter().zip_eq(&col_indices).map(|(row, col)| *row * col).collect();
+    cfg_iter_mut!(row_col_vals).zip_eq(&row_col_indices).for_each(|(v, rc)| *v *= rc);
 
-    cfg_iter_mut!(val_vec).zip_eq(inverses).for_each(|(v, inv)| *v *= inv);
     end_timer!(lde_evals_time);
 
+    // Fill up the evaluations to the next power of two
     for _ in count..non_zero_domain.size() {
-        col_vec.push(elems[0]);
-        row_vec.push(elems[0]);
-        val_vec.push(F::zero());
+        row_indices.push(F::one());
+        col_indices.push(F::one());
+        row_col_indices.push(F::one());
+        row_col_vals.push(F::zero());
     }
```

Comment:
Let's do the padding before we construct `row_col_indices`, as the additional pushes currently cause `row_col_indices` to reallocate

Reply:
We should also avoid superfluous multiplication of row and col. Tried another approach here which first allocates the full vector, then only multiplies a (sub)slice of it: 92ec45d61